py/persipubsub/queue.py

```diff
@@ -0,0 +1,489 @@
+            # check if database is not empty
+            if cursor.first():
+                key = cursor.key()
+                cursor.pop(key=key)
+                pending_value = txn.get(key=key, db=pending_db)
+                pending_num = persipubsub.bytes_to_int(pending_value)
+                decreased_pending_num = pending_num - 1
+                assert decreased_pending_num >= 0
+                txn.put(
+                    key=key,
+                    value=persipubsub.int_to_bytes(decreased_pending_num),
+                    db=pending_db)
+            else:
+                raise RuntimeError("No message to pop")
+
+    def prune_dangling_messages(self) -> None:
+        """
+        Prune dangling messages in the queue.
+
+        Definition of dangling messages:
+        - having no pending subscribers
+        - exists longer than timeout allows
+        :return:
+        """
+        assert isinstance(self.subscriber_ids, List)
+        _prune_dangling_messages_for(
+            queue=self, subscriber_ids=self.subscriber_ids)
+
+    def check_current_lmdb_size(self) -> int:
+        """
+        Check current LMDB size in bytes.
+
+        Check size of data database by approximating size with multiplying page
+        size with number of pages.
+
+        :return: data database size in bytes
+        """
+        lmdb_size_bytes = 0
+        assert isinstance(self.env, lmdb.Environment)
+        with self.env.begin(write=False) as txn:
+            data_db = self.env.open_db(
+                key=persipubsub.DATA_DB, txn=txn, create=False)
+            data_stat = txn.stat(db=data_db)
+            lmdb_size_bytes += data_stat['psize'] * (
+                data_stat['branch_pages'] + data_stat['leaf_pages'] +
+                data_stat['overflow_pages'])
+
+        return lmdb_size_bytes
+
+    def count_msgs(self) -> int:
+        """
+        Count number of messages in database.
+
+        Count number of messages stored in named database 'meta_db'.
+
+        :return: number of messages in database
+        """
+        assert isinstance(self.env, lmdb.Environment)
+        with self.env.begin(write=False) as txn:
+            meta_db = self.env.open_db(
+                key=persipubsub.META_DB, txn=txn, create=False)
+            meta_stat = txn.stat(db=meta_db)  # type: Dict[str, int]
+
+        return meta_stat['entries']
+
+    def vacuum(self) -> None:
+        """
+        Clean database when needed.
+
+        :return:
+        """
+        assert isinstance(self.hwm, HighWaterMark)
+        self.prune_dangling_messages()
+        msgs_num = self.count_msgs()
+        if msgs_num >= self.hwm.max_msgs_num:
+            self.prune_messages()
+        lmdb_size_bytes = self.check_current_lmdb_size()
+        if lmdb_size_bytes >= self.hwm.hwm_lmdb_size_bytes:
+            self.prune_messages()
+
+    def prune_messages(self) -> None:
+        """
+        Prune one half of the messages stored.
```

Comment:
Why do you prune a half of the messages? Please document this in more detail. Also, rename to `prune_half_of_messages`.